README.md
# Theremin Remote (a.k.a. HAKeyboardRemote)
A simple system for using HID (human input devices, e.g. mice, keyboards, etc.) as remote controls for issuing remote API requests. This is a configurable Ruby script that allows Linux users to use devices (presumably wireless) as dimmer switches or whatever else you might have in mind, mainly for home automation. 

The project is nicknamed "Theremin" because, as it stands now, you can use the touchpad to control home lights much the way a trained musician plays a theremin (except that you do have to touch the touchpad). Anyway, it maps the X & Y coordinates of the touchpad into the HSL or XY color space (aka CIE 1931), so you wind up with a sort of two-dimensional dimmer switch. Actually using the thing this way is kind of a nightmare so just use the buttons.

This is a work in progress. The purpose was to take a $15 miniature keyboard-and-trackpad combo, tie it into HA-Bridge and/or OpenHAB2, and mount it on the wall near my front door to control all my "smart lights" from one place without having to talk out loud to my furniture (e.g. Alexa). I also wanted to use the mini-keyboard's built-in trackpad as a color selector for my entire apartment, which consists of various different brands of lights (a major factor in the need for HA-Bridge and OpenHAB2 et al.). So I've mounted the controller by my door using magnets (so I can easily remove it to use on the couch, or recharge it) and now I can turn everything on or off when I leave or enter my apartment.

## Working:
- button mapping
- curl commands issued to HA-Bridge or Openhab
- using Trackpad to control XY-colorspace 
- color randomization and rotation
- multiple input devices
- direct connection support for HA-bridge or directly to OpenHAB2 (the second seems faster, more reliable, and generally more color-intense!)
- custom request/toggle commands (acting against a received state) - this is a little primitive for now, but works ok with Openhab
- arbitrary commands ("exec" -- I use this to run [advanced eISCP commands via Python](https://github.com/miracle2k/onkyo-eiscp) on my Onkyo stereo, e.g. for FM and Internet Radio presets, since OpenHAB is not supporting them yet)

## Security warning:
- The "exec" and "eval" commands in my config script could allow arbitrary code execution if for some weird reason you were to run a config file you didn't write yourself. I can't imagine anyone doing that, but thought I should mention it.

## Roadmap:
- user guide (if people are actually interested in Theremin Remote at all)
- scenes and what-not
- whatever other improvements I think of
- fun programming routines for the lights (make 'em dance)

Most such changes are trivial modifications to the Ruby code.

## Contributing:
I'll be very happy if you submit PRs with more functionality to build off this base. It's not a complicated piece of code; I only wrote this because it didn't seem to exist already.

## Prerequisites:
- You probably won't find much use for this without at least one external keyboard and/or mouse. 
- You might need a rudimentary understanding of the /dev/input devices and input event handling in a linux environment. Well honestly, you can probably figure most of what you need using my config.json as an example.
- I have no idea if this will work in non-Linux environment... it would probably work on a Mac, but I doubt it would work under Windows. It has been tested very successfully using OpenHABian on a Raspberry Pi Model 3 B.
- You need some sort of target for the requests. My code is based on a modified HA-Bridge-- modified in that I added HSL colorspace support, which HA-Bridge is supposed to pass onto OpenHAB. At the time of this writing it is an open pull request at ha-bridge, awaiting review... You can just pull it in and compile it locally. https://github.com/bwssytems/ha-bridge/pull/1028 and then you can use ${color.hsl} in your colorization requests to HA-Bridge. HOWEVER, as of 12/19/2018 I've added support for OpenHAB2 directly (though it is a work-in-progress) which seems better overall, and obviates any need for the CIE 1931 / a.k.a. XY colorspace which is just a nightmare to handle anyway.

## Suggested hardware:
- [Mitid Wireless Mini Keyboard](https://www.amazon.com/Mitid-Wireless-Keyboard-Touchpad-Raspberry/dp/B01E565PIQ/ref=as_li_ss_tl?ie=UTF8&tag=skullbasher-20&linkCode=as2&camp=217145&creative=399373&creativeASIN=1608870243) - I suggest this one because it appears to have a nice HSL bar printed right on it. I don't actually have this one; mine is made by Inland and has a few more buttons and back lighting. Any keyboard and/or mouse should work (they do not have to be wireless).
- Some cheap computer/Raspberry Pi/whatever that can handle OpenHAB and/or HA-Bridge for running the remote listener.

## Ruby and Ruby Prerequisites: 
You will need to install Ruby 2.0 or greater and the ruby2.3-dev and libcurl-dev libs:
```
sudo apt-get install ruby2.3 ruby2.3-dev libcurl4-openssl-dev
```
then:
```
gem install curb
gem install ffi
gem install device_input
gem install json
```

Copy the config.json.example to config.json, and modify it to your Openhab2 and/or HABridge settings before you run the script. The mode option must be 'openhab', anything else is currently treated as 'ha_bridge'. If using the ha_bridge mode, light IDs should correspond to those in your HA-Bridge. Otherwise you must also set mappings for the lights in the section called 'openhab_devices' to correspond to the IDs for your commands. Someday maybe I will add a GUI.

## A bit of miscellany:
- I've included a simple script, internet_radio_preset.sh, which simply writes a counter so you can step through internet radio presets on devices that don't support it (e.g. the Onkyo NR-626). It is really somewhat out-of-scope for this project, but it hardly merits its own repo either. If you have an Onkyo receiver you can use this with [onkyo-eiscp](https://github.com/miracle2k/onkyo-eiscp).

## Setup:
- Copy the config.json.example file to config.json. Mostly it should be self-explanatory. The most important thing is to get the device ID of your mouse and keyboard correctly, and the light IDs on HA-bridge. Depending on your set-up the username may need to change, and depending on your setup, the ha_bridge_url value. If you have the keyboard/mouse on the same machine that's running HA-Bridge you should be fine with "localhost". 
- Your results will probably be screwy if you include switches instead of color bulbs in commands that rely on color (e.g. random, rotate, and the like), or color assignments to dimmer lights. 
- Physically you'll want to check the range of your remote and make sure that whatever PC/Raspberry/etc. is running this script is within range of your remote keyboard and has connectivity to your ha-bridge server.
- Choose a "mode" value in your config.json. The options are "openhab" and "ha_bridge". In "openhab" mode, you must define device IDs in the openhab_devices section of the config.json. In "ha_bridge" mode the IDs are the ones ha_bridge itself uses, so it's somewhat simpler to set up the config (though setting up HA_Bridge itself is an equal chore). Openhab mode is slightly faster since it eliminates the middleman, and it lets you use HSL colors which are much easier to work with than that whole XY/CIE1931 thing. But since I started coding this through ha_bridge before realizing any of that, I've left both modes available.

## Run:

```
ruby theremin.rb 
```




CHANGELOG.md
Version 0.27
- Add ability to filter out devices (different actions for specific devices by path)

Version 0.26
- Add ability to chain together multiple URLs

Version 0.25
- Add Echo pause/unpause support

Version 0.24
- improve everything

Version 0.2
- Add stereo support
- Add "openhab" mode
- Add HSL mode for trackpad/mouse

Version 0.1
- Initial stuff
- Everything working in "ha_bridge" mode


ruby_remote.rb
require 'device_input'
require 'manticore'
require 'json'

config_file = File.read('config.json')
CONFIGURATION = JSON.parse(config_file)
BRIGHTNESS_SYNC_PERIOD = 20
MAX_CONCURRENT_REQUESTS = 8
REPEAT_DELAY = 20
DEBUG = !!CONFIGURATION['debug']

# use two clients because the MOUSE client should wipe out any existing connections each time it sends
CLIENTS = {
  'keyboard' => Manticore::Client.new(request_timeout: 3, connect_timeout: 5, socket_timeout: 4, pool_max: 15, pool_max_per_route: 2),
  'mouse' => Manticore::Client.new(request_timeout: 3, connect_timeout: 5, socket_timeout: 4, pool_max: 15, pool_max_per_route: 2)
}

@px = 0.0
@py = 0.0
@brightness ||= {} 
@color ||= {} 

def mapped_keys
  @mapped_keys ||= CONFIGURATION['keyboard_bindings'].flat_map { |mapping| mapping['key'] }
end

def get_state
  result = Manticore.get("#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights")
  JSON.parse(result.body).each do |k,v|
    @color[k.to_i] = v['state']['xy']
    @brightness[k.to_i] = v['state']['bri']
  end
end

def make_dimmer_call(light, intensity)
  ["#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights/#{light}/state", { body: { 'bri': intensity }.to_json }]
end

def make_color_call(light, values)
  ["#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights/#{light}/state", { body: { 'xy': values }.to_json }]
end

def make_power_call(light, on_state)
  @brightness[light] = 0 if !on_state
  ["#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights/#{light}/state", { body: { 'on': on_state }.to_json }]
end

def parallel_calls(datas, client_key)
  if client_key == 'mouse'
    CLIENTS[client_key].clear_pending
  end
  datas.each do |data|
    response = CLIENTS[client_key].background.parallel.put(*data)
    response.on_success do |response|
      puts "SUCCESS: #{response}" if DEBUG 
    end.on_failure do |response|
      puts "FAILED: #{response}" if DEBUG 
    end
  end
  CLIENTS[client_key].execute! 
rescue => e
  puts e.message if DEBUG
end

def perform_action_for(key, code = 0)
  bindings = CONFIGURATION['keyboard_bindings'].select { |item| item['key'] == key || (item['key'].is_a?(Array) && item['key'].include?(key)) }
  keyboard_actions = []
  bindings.each do |binding|
    if code == 0 # binding['repeatable'] # TODO: repeating requires awareness of when the key goes down and when up
      if binding['type'] == 'ha-bridge' 
        if binding['action'] == 'on' || binding['action'] == 'off' 
          keyboard_actions = binding['lights'].collect { |light| make_power_call(light, binding['action'] == 'on') } 
        elsif binding['action'] == 'dim'
          keyboard_actions = binding['lights'].collect do |light| 
            @brightness[light] += binding['value']
            @brightness[light] = min_max(@brightness[light], 0, 254)
            make_dimmer_call(light, @brightness[light].to_i)
          end 
	elsif binding['action'] == 'dim_multiply'
          keyboard_actions = binding['lights'].collect do |light| 
            @brightness[light] *= (binding['value'] || 0.5)
            @brightness[light] = min_max(@brightness[light], (binding['value'].to_f >= 1 ? 16 : 0), 254) # if binding-value is > 1, we mean to raise the brightness, so don't let the outcome be 0
            make_dimmer_call(light, @brightness[light].to_i)
          end 
        elsif binding['action'] == 'random'
          keyboard_actions = binding['lights'].collect { |light| make_color_call(light, [rand(), rand()]) }
        elsif binding['action'] == 'rotate'
          target_lights = (binding['reversed'] ? binding['lights'].reverse : binding['lights'])
          get_state  
          color = @color[target_lights.last] 
          keyboard_actions = target_lights.collect do |light| 
            new_color_xy = color
            color = @color[light]
	    puts "#{light}: #{new_color_xy} / old color #{color}"
	    @color[light] = new_color_xy
            make_color_call(light, new_color_xy)
          end
        end
      end
    end
  end
  keyboard_actions
end

def min_max(val,min,max) 
  a = val < min ? min : val
  a > max ? max : a
end

def mouse_input_monitor
  File.open(CONFIGURATION['mouse_device'], 'r') do |dev|
    DeviceInput.read_loop(dev) do |event|
      puts "Moved by #{event.data.value.to_f}" if (event.code == "X" || event.code == "Y") if DEBUG 
      if (['X','Y'].include?(event.code))
        get_state

        calls = CONFIGURATION['mouse_binding']['lights'].collect do |light| 
	  new_x = new_y = 0
          if (event.code == "X")
            new_x = @color[light][0] + (event.data.value.to_f / CONFIGURATION['max_x_resolution'])
	    new_y = @color[light][1]
          end
          if (event.code == "Y")
	    new_x = @color[light][0]
            new_y = @color[light][1] - (event.data.value.to_f / CONFIGURATION['max_y_resolution']) # the operator is a - because Y is inverted (0,0 is the top left corner)
          end
	  if CONFIGURATION['drift']
	    new_x += (rand() * CONFIGURATION['drift'] * 2) - CONFIGURATION['drift']
	    new_y += (rand() * CONFIGURATION['drift'] * 2) - CONFIGURATION['drift']
          end 
          new_xy = [min_max(new_x,0,1.0),min_max(new_y,0,1.0)] 
	  @color[light] = new_xy 
          make_color_call(light, new_xy)
        end
        parallel_calls(calls, 'mouse')
      end
    end
  end
end

def keyboard_input_monitor
  File.open(CONFIGURATION['keyboard_device'], 'r') do |dev|
    DeviceInput.read_loop(dev) do |event|
      if event.type == 'EV_KEY' 
        if mapped_keys.include?(event.code)  
          parallel_calls(perform_action_for(event.code, event.data.value), 'keyboard')
        else
          puts "#{event.code} is not a mapped key" if DEBUG
        end
      end
    end
  end
end

get_state
pp @brightness if DEBUG
puts "Bound keys: #{mapped_keys.join(",")}" if DEBUG
t3 = Thread.new do 
  loop do
    begin
      mouse_input_monitor
    rescue => e
      pp e
    end
  end
end unless CONFIGURATION['mouse_disabled']
t4 = Thread.new do 
  loop do
    begin
      keyboard_input_monitor
    rescue => e
      pp e
    end
  end
end unless CONFIGURATION['keyboard_disabled']
t3.join unless CONFIGURATION['mouse_disabled']
t4.join unless CONFIGURATION['keyboard_disabled']


internet_radio_preset.sh
#!/bin/bash
# Script for stepping through Internet radio presets (or any similar counter) on an Onkyo-626 which does not have this functionality.
# MAX should be set to the highest number preset you have, and MIN should probably be 1.  
FILENAME=internet_radio_preset.txt
MAX=32
MIN=1

if [ -e $FILENAME ]
then
  CURRENT_VALUE=`cat $FILENAME`
else
  CURRENT_VALUE=0
fi
  
if [ " $1" == " down" ]
then
  let CURRENT_VALUE=$(($CURRENT_VALUE - 1))
else
  let CURRENT_VALUE=$(($CURRENT_VALUE + 1))
fi

re='^ [0-9]+$'
if [[ " $1" =~ $re ]] ; then
   let CURRENT_VALUE=$1
fi

if [ $CURRENT_VALUE -lt $MIN ]
then
  let CURRENT_VALUE=$MAX
fi

if [ $CURRENT_VALUE -gt $MAX ]
then
  let CURRENT_VALUE=$MIN
fi

echo $CURRENT_VALUE > $FILENAME
echo `printf '%02x\n' $CURRENT_VALUE` # because the Onkyo requires hex digits for some reason


theremin_remote.rb
require 'rubygems'
require 'device_input'
require 'curb'
require 'json'

require 'ffi'
require 'ffi/tools/const_generator'

VERSION = 0.27

config_file = File.read('config.json')
CONFIGURATION = JSON.parse(config_file)
BRIGHTNESS_SYNC_PERIOD = 20
MAX_CONCURRENT_REQUESTS = 14
REPEAT_DELAY = 20
DEBUG = !!CONFIGURATION['debug']
DEVICES = CONFIGURATION['devices']
MODE = CONFIGURATION['mode'] # "openhab" or "ha_bridge" 

HUE_RANGE = 360 # can't imagine this changing
SATURATION_RANGE = 100
MAX_BRIGHTNESS = (MODE == 'openhab' ? 100 : 255)

# curl options
CURL_EASY_OPTIONS = {:follow_location => true}
CURL_MULTI_OPTIONS = {:pipeline => Curl::CURLPIPE_HTTP1}

# for openhab HSB mapping
HUE = 0
SATURATION = 1
BRIGHTNESS = 2

# get the EVIOCGRAB constant
cg = FFI::ConstGenerator.new('input') do |gen|
  gen.include('linux/input.h')
  gen.const(:EVIOCGRAB, '%u', '(unsigned)')
end
EVIOCGRAB = cg['EVIOCGRAB'].to_i


# use two clients because the MOUSE client should wipe out any existing connections each time it sends
# this probably needs to go
CLIENTS = {
    'keyboard' => Curl::Multi,
    'mouse' => Curl::Multi
}

@px = 0.0
@py = 0.0
@state = {}

def mapped_keys
  @mapped_keys ||= CONFIGURATION['keyboard_bindings'].flat_map { |mapping| mapping['key'] }
end

class String
  def from_hsl
    self.split(',') # maps to HUE,SATURATION,BRIGHTNESS
  end
end

class Array
  def to_hsl
    self.map(&:to_i).join(',') # maps to "HUE,SATURATION,BRIGHTNESS"
  end
end

def random_color(light)
  if MODE == 'openhab'
    random_color = [(rand() * HUE_RANGE).to_i,
                    (rand() * SATURATION_RANGE).to_i,
                    brightness_of(light)].to_hsl
  else
    random_color = [rand(), rand()]
  end
end

def openhab_device_map_for(items)
  CONFIGURATION['openhab_devices'].select { |dev| dev } # TODO 
end

def color_of(key)
  if MODE == 'openhab'
    @state.dig(key.to_i, 'color', HUE).to_i
  else
    @state.dig(key.to_i, 'xy') || 0
  end
end

def all_values_of(key)
  if MODE == 'openhab'
    @state.dig(key.to_i, 'color')
  else
    @state.dig(key.to_i, 'xy') || 0
  end
end

def set_values(key, values)
  @state[key.to_i] ||= {}
  if MODE == 'openhab'
    @state[key.to_i]['color'] = values
  else
    @state[key.to_i]['xy'] = values
  end
end

def saturation_of(key)
  if MODE == 'openhab'
    @state.dig(key.to_i, 'color', SATURATION).to_i
  else
    MAX_SATURATION
  end
end

def brightness_of(key)
  if MODE == 'openhab'
    @state.dig(key.to_i, 'color', BRIGHTNESS).to_i
  else
    @state.dig(key.to_i, 'bri') || MAX_BRIGHTNESS
  end
end

def set_on_off(key, val)
  @state[key.to_i] ||= {}
  @state[key.to_i]['on'] = !!val
end

def is_on?(key)
  puts @state.to_json
  @state.dig(key.to_i, 'on') || false
end

def is_off?(key)
  !is_on?(key)
end

def has_color?(key)
  @state.has_key?(key) && (@state[key].has_key?('xy') || @state[key].has_key?('color'))
end

def set_color(key, value)
  @state[key.to_i] ||= {}
  if MODE == 'openhab'
    @state[key.to_i]['color'] ||= []
    @state[key.to_i]['color'][HUE] = value
  else
    @state[key.to_i]['xy'] = value if value.is_a?(Array) && value.length == 2 && value.all? { |v| v.is_a?(Float) }
  end
end

def set_saturation(key, value)
  @state[key.to_i] ||= {}
  if MODE == 'openhab'
    @state[key.to_i]['color'] ||= []
    @state[key.to_i]['color'][SATURATION] = value.to_i
  else
    @state[key.to_i]['bri'] = value.to_i
  end
end

def set_brightness(key, value)
  @state[key.to_i] ||= {}
  if MODE == 'openhab'
    @state[key.to_i]['color'] ||= []
    @state[key.to_i]['color'][BRIGHTNESS] = value.to_i
  else
    @state[key.to_i]['bri'] = value.to_i
  end
end

def get_state
  puts "Getting state..." if DEBUG
  if MODE == 'openhab'
    CONFIGURATION['openhab_devices'].each do |device|
      result = Curl::Easy.perform("#{CONFIGURATION['openhab_url']}/rest/items/#{device['name']}/state")
      @state[device['id']] ||= {}
      if device['type'] == 'color'
        tcolor = result.body.split(",")
        @state[device['id']].merge!({'color' => tcolor})
        @state[device['id']]['on'] = (@state[device['id']]['color'][BRIGHTNESS].to_i != 0)
      end
    end
  else
    result = Curl::Easy.perform("#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights")
    JSON.parse(result.body).each do |k, v|
      @state[k.to_i] = v['state']
    end
  end
end

def openhab_mapping_for(light)
  # provide the label for openhab lights, for the URL, otherwise return self
  CONFIGURATION['openhab_devices'].detect { |dev| dev['id'] == light }['name']
end

def type_of(light)
  # provide the label for openhab lights, for the URL, otherwise return self
  CONFIGURATION['openhab_devices'].detect { |dev| dev['id'] == light }['type']
end

def make_dimmer_call(light, intensity)
  if MODE == 'openhab'
    {url: "#{CONFIGURATION['openhab_url']}/rest/items/#{openhab_mapping_for(light)}",
     post_fields: [color_of(light), saturation_of(light), intensity].to_hsl}
  else
    {url: "#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights/#{light}/state",
     post_fields: {'bri' => intensity}.to_json}
  end
rescue
  nil
end

def make_color_call(light, values)
  if MODE == 'openhab'
    {url: "#{CONFIGURATION['openhab_url']}/rest/items/#{openhab_mapping_for(light)}",
     post_fields: values}
  else
    {url: "#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights/#{light}/state",
     post_fields: {'xy': values}.to_json}
  end
rescue
  nil
end

def make_power_call(light, on_state)
  if MODE == 'openhab'
    if type_of(light) == 'echo'
      {url: "#{CONFIGURATION['openhab_url']}/rest/items/#{openhab_mapping_for(light)}",
       post_fields: (on_state ? "PLAY" : "PAUSE")}
    else
      {url: "#{CONFIGURATION['openhab_url']}/rest/items/#{openhab_mapping_for(light)}",
       post_fields: (on_state ? "ON" : "OFF")}
    end
  else
    set_brightness(light, 0) if !on_state
    {url: "#{CONFIGURATION['ha_bridge_url']}/api/#{CONFIGURATION['ha_bridge_username']}/lights/#{light}/state",
     post_fields: {'on': on_state}.to_json}
  end
rescue
  nil
end

def parallel_calls(datas, client_key)
  if datas
    m = Curl::Multi.new
    datas.flatten.compact.each do |data|
      puts "Curling #{data.class}" if DEBUG
      c = Curl::Easy.http_post(data[:url]) do |curl|
        curl.method('post')
        curl.follow_location = true
        curl.post_body = data[:post_fields]
        curl.headers['Content-Type'] = 'text/plain'
      end
      m.add(c)
    end

    m.perform
  end
end

def action_item(binding, device_name)
  
  if (binding['except'].nil? || !binding['except'].to_a.include?(device_name)) && (binding['only'].nil? || binding['only'].to_a.include?(device_name))
  puts "Executing: #{binding['name']} from #{device_name}" if binding['name'] && DEBUG

  if binding['action'] == 'reroute'
    perform_action_for(binding['substitute'], 0, device_name)
  elsif binding['action'] == 'on' || binding['action'] == 'off'
    get_state
    binding['lights'].map do |light|
      set_on_off(light, binding['action'] == 'on')
      make_power_call(light, binding['action'] == 'on')
    end
  elsif binding['action'] == 'dim'
    get_state
    binding['lights'].map do |light|
      t_bri = brightness_of(light) + (binding['value'] || 0.5)
      set_brightness(light, min_max(t_bri, 0, MAX_BRIGHTNESS))
      make_dimmer_call(light, brightness_of(light).to_i)
    end
  elsif binding['action'] == 'dim_multiply'
    get_state
    binding['lights'].map do |light|
      t_bri = brightness_of(light) * (binding['value'] || 0.5)
      set_brightness(light, min_max(t_bri, (binding['value'].to_f >= 1 ? 16 : 0), MAX_BRIGHTNESS)) # if binding-value is > 1, we mean to raise the brightness, so never let the outcome be 0
      make_dimmer_call(light, brightness_of(light))
    end
  elsif binding['action'] == 'random'
    get_state
    binding['lights'].map { |light| make_color_call(light, random_color(light)) }
  elsif binding['action'] == 'white' || binding['action'] == 'color'
    if MODE == 'ha_bridge'
      col = [0.33333333333, 0.33333333333]
      if binding['action'] == 'color'
        col[0] = binding['x'] if binding['x']
        col[1] = binding['y'] if binding['y']
      end
    elsif MODE == 'openhab'
      col = "0,0,100"
    end

    get_state
    binding['lights'].map do |light|
      if has_color?(light) && MODE == 'openhab' && type_of(light) == 'color'
        if binding['action'] == 'white'
          col = "0,0,100"
        else
          h = binding['h'] ? binding['h'] : color_of(light)
          s = binding['s'] ? binding['s'] : saturation_of(light)
          l = binding['l'] ? binding['l'] : brightness_of(light)
          col = [h, s, l].to_hsl
        end
        make_color_call(light, col)
      elsif binding['switches_on']
        make_power_call(light, true)
      end
    end.compact
  elsif binding['action'] == 'toggle'
    # Generally you wouldn't want to actually toggle everything, but rather set them all to the same thing.
    # So, this determines first if any of the lights are currently on, and if so, it turns them off.
    # Otherwise it turns them all on.
    get_state
    anything_on = binding['lights'].any? { |light| is_on?(light) }
    binding['lights'].each { |light| set_on_off(light, !anything_on) }
    binding['lights'].map { |light| make_power_call(light, !anything_on) }
  elsif binding['action'] == 'rotate'
    get_state
    target_lights = (binding['reversed'] ? binding['lights'].reverse : binding['lights'])
    values = all_values_of(target_lights.last)
    target_lights.map do |light|
      new_values = values
      values = all_values_of(light)
      set_values(light, new_values)
      make_color_call(light, MODE == 'openhab' ? new_values.to_hsl : new_values)
    end
  elsif binding['action'] == 'custom'
    new_state = binding['body']
    if binding['state_url']
      puts "Custom action #{binding['eval']}" if DEBUG
      state = JSON.parse(Curl.send((binding['state_method'] || 'get').to_sym, binding['state_url']).body)[(binding['state_variable'] || 'state')]
      puts "Current state is #{state}" if DEBUG
      new_state = eval(binding['eval'])
    end
    if binding['url'].is_a?(String)
      target = [binding['url']]
    else
      target = binding['url']
    end
    target.each { |url|  Curl.send((binding['method'] || 'post').to_sym, url, (binding['body'] || new_state).to_s) { |c| c.headers['Content-Type'] = 'text/plain' } }
    nil
  elsif binding['action'] == 'exec'
    system(item['exec'])
    nil
  elsif binding['chain']
    binding['chain'].map { |chain| action_item(chain, device_name) }
  end.to_a.compact
  end
end

def perform_action_for(key, code = 0, device_name)
  bindings = CONFIGURATION['keyboard_bindings'].select { |item| item['key'] == key || (item['key'].is_a?(Array) && item['key'].include?(key)) }
  bindings.flat_map { |binding| action_item(binding, device_name) if code == 0 }.compact
end

def min_max(val, min, max)
  a = val < min ? min : val
  a > max ? max : a
end

def mouse_xy(event)
  get_state
  CONFIGURATION['mouse_binding']['lights'].map do |light|
    new_x = new_y = 0
    if (event.code == "X")
      new_x = color_of(light)[0] + (event.data.value.to_f / CONFIGURATION['max_x_resolution'])
      new_y = color_of(light)[1]
    end
    if (event.code == "Y")
      new_x = color_of(light)[0]
      new_y = color_of(light)[1] - (event.data.value.to_f / CONFIGURATION['max_y_resolution']) # the operator is a - because Y is inverted (0,0 is the top left corner)
    end
    if CONFIGURATION['drift']
      new_x += (rand() * CONFIGURATION['drift'] * 2) - CONFIGURATION['drift']
      new_y += (rand() * CONFIGURATION['drift'] * 2) - CONFIGURATION['drift']
    end
    new_xy = [min_max(new_x, 0, 1.0).to_i, min_max(new_y, 0, 1.0).to_i]
    set_color(light, new_xy)
    make_color_call(light, new_xy)
  end
end

def mouse_hsl(event)
  get_state
  CONFIGURATION['mouse_binding']['lights'].map do |light|
    new_h = new_s = 0
    if (event.code == "X")
      new_h = color_of(light) + 360 * (event.data.value.to_f / CONFIGURATION['max_x_resolution'])
      new_s = saturation_of(light)
    end
    if (event.code == "Y")
      new_h = color_of(light)
      new_s = saturation_of(light) + 100 * (event.data.value.to_f / CONFIGURATION['max_y_resolution'])
    end
    if CONFIGURATION['drift']
      new_h += (rand() * CONFIGURATION['drift'] * 2) - CONFIGURATION['drift']
      new_s += (rand() * CONFIGURATION['drift'] * 2) - CONFIGURATION['drift']
    end
    new_hue = min_max(new_h, 0, 359.99)
    new_saturation = min_max(new_s, 0, 100.0)
    new_brightness = brightness_of(light)

    set_color(light, new_hue)
    set_saturation(light, new_saturation)
    make_color_call(light, [new_hue, new_saturation, new_brightness].to_hsl)
  end
end

def input_monitor(device)
  if File.readable?(device)
    dev = File.new(device)
    begin
      dev.ioctl(EVIOCGRAB, 1) # prevent other things from listening to the device
    rescue
    end
    DeviceInput.read_loop(dev) do |event|
      puts "Received #{event}" if DEBUG
      if event.type == 'EV_KEY' && !CONFIGURATION['keyboard_disabled'] # KEYBOARD DEVICES
        puts "Received: #{event.code}" if DEBUG
        if mapped_keys.include?(event.code)
          calls = parallel_calls(perform_action_for(event.code, event.data.value, device), 'keyboard')
        else
          puts "#{event.code} is not a mapped key" if DEBUG
        end
      elsif !CONFIGURATION['mouse_disabled']
        puts "Moved by #{event.data.value.to_f}" if (event.code == "X" || event.code == "Y") if DEBUG
        puts "Mouse action: #{event}" if DEBUG

        if event.code == 'Wheel'
          parallel_calls(perform_action_for("Wheel#{event.data.value == -1 ? "Down" : "Up"}", 0, device), 'keyboard')
        elsif !CONFIGURATION['mouse_movements_disabled'] && (['X', 'Y'].include?(event.code))  
          if CONFIGURATION['mouse_effect'] == 'xy'
            calls = mouse_xy(event)
          elsif CONFIGURATION['mouse_effect'] != 'none'
            calls = mouse_hsl(event)
          end
        end
        parallel_calls(calls, 'mouse')
      end
    end
  else
    puts "Device #{device} could not be read, skipping"
    false
  end
end

def start
  puts "Bound keys: #{mapped_keys.join(",")}" if DEBUG
  threads = []
  DEVICES.each_index do |i|
    threads.push(Thread.new(i) do
      puts "Adding device #{DEVICES[i]}" if DEBUG
      loop do
        begin
          break if !input_monitor(DEVICES[i])
        rescue => e
          puts "** ERROR! #{e.message} **"
          puts e.backtrace.join("\n")
          puts "** END OF ERROR **"
        end
      end
    end)
  end
  DEVICES.each_index { |i| threads[i].join }
end

start
